### PR TITLE
#830 fix bug

### LINKED ISSUE #830: SBML parser: incomplete reaction equations
SBML parser appears to ignore terms in differential equations for products of reactions. E.g.

```
<?xml version="1.0" encoding="UTF-8"?>
<sbml xmlns="http://www.sbml.org/sbml/level3/version2/core" level="3" version="2">
  <model id="test" timeUnits="second">

    <listOfUnitDefinitions>
      <unitDefinition id="per_second">
        <listOfUnits>
          <unit kind="second" exponent="-1" scale="0" multiplier="1"/>
        </listOfUnits>
      </unitDefinition>
    </listOfUnitDefinitions>

    <listOfCompartments>
      <compartment id="central" name="central" size="1" units="liter"/>
    </listOfCompartments>

    <listOfSpecies>
      <!-- Coagulation factor XII -->
      <species id="sp1" name="sp1" compartment="central" initialAmount="0" hasSubstanceUnits="false" substanceUnits="mole"/>
      <species id="sp2" name="sp2" compartment="central" initialAmount="0" hasSubstanceUnits="false" substanceUnits="mole"/>
    </listOfSpecies>

    <listOfParameters>
      <parameter id="rate1" value="1" constant="true" units="per_second"/>
      <parameter id="rate2" value="1" constant="true" units="per_second"/>
    </listOfParameters>

    <listOfReactions>
      <reaction id="reaction1" name="reaction1" reversible="false" fast="false">
        <listOfReactants>
          <speciesReference species="sp1"/>
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="sp2"/>
        </listOfProducts>
        <kineticLaw>
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <times/>
              <ci> central </ci>
              <ci> rate1 </ci>
              <ci> sp1 </ci>
            </apply>
          </math>
        </kineticLaw>
      </reaction>

      <reaction id="reaction2" name="reaction2" reversible="false" fast="false">
        <listOfReactants>
          <speciesReference species="sp1"/>
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="sp2"/>
        </listOfProducts>
        <kineticLaw>
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <times/>
              <ci> central </ci>
              <ci> rate2 </ci>
              <ci> sp1 </ci>
            </apply>
          </math>
        </kineticLaw>
      </reaction>
    </listOfReactions>
  </model>
</sbml>
```

leads to the following system of ODEs

```
[[model]]
name: test
# Initial values
central.sp1_amount = 0
central.sp2_amount = 0

[central]
size = 1
    in [L]
dot(sp1_amount) = -(size * myokit.rate2 * sp1_concentration)
    in [mol]
sp1_concentration = sp1_amount / size
    in [M]
dot(sp2_amount) = size * myokit.rate2 * sp1_concentration
    in [mol]
sp2_concentration = sp2_amount / size
    in [M]

[myokit]
rate1 = 1
    in [S/F]
rate2 = 1
    in [S/F]
time = 0 [s] bind time
    in [s]
```

when it should create a model that looks like

```
[[model]]
name: test
# Initial values
central.sp1_amount = 0
central.sp2_amount = 0

[central]
size = 1
    in [L]
dot(sp1_amount) = -(size * myokit.rate1 * sp1_concentration) - size * myokit.rate2 * sp1_concentration
    in [mol]
sp1_concentration = sp1_amount / size
    in [M]
dot(sp2_amount) = size * myokit.rate1 * sp1_concentration + size * myokit.rate2 * sp1_concentration
    in [mol]
sp2_concentration = sp2_amount / size
    in [M]

[myokit]
rate1 = 1
    in [S/F]
rate2 = 1
    in [S/F]
time = 0 [s] bind time
    in [s]
```

### PR DESCRIPTION
**Pull request by DavAug**
Closes #830 .

### CODE REVIEW

**Comment by MichaelClerx** [comment]
Great catch, and thanks for the fix!

>>> if var.rhs().eval() or not var.rhs().is_literal():

I'd probably flip these around because the second check is cheaper (and we can avoid evaluating an rhs that depends on other vars it it's true)

Could you also add a line to the changelog?

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ This page lists the main changes made to Myokit in each release.
   - [#821](https://github.com/MichaelClerx/myokit/pull/821) Removed the method `myokit.Sundials.version_int`.
 - Fixed
   - [#805](https://github.com/MichaelClerx/myokit/pull/805) Added missing command line utilities to the documentation, and re-organised the command-line tool docs.
+  - [#831](https://github.com/MichaelClerx/myokit/pull/831) Fixed bug of SBML parser when parsing reactions.
```
Comment (MichaelClerx) [inline CHANGELOG.md:27]:
```suggestion
  - [#831](https://github.com/MichaelClerx/myokit/pull/831) Fixed bug in SBML parser that caused it to ignore terms in differential equations for products of reactions.
```

**Comment by MichaelClerx** [comment]
Could we also have a test that produces the error without this fix? And a test to show that 0s are omitted if possible (if we don't already have that)

**Comment by DavAug** [comment]
Sure no problem! Would you test the equations for the myokit model directly? E.g. by asserting that the model.code() returns a mmt file that we expect? So far we mostly tested that the simulation results are the same as solutions from the SBML test suite but that wouldn’t show whether the 0‘s are present in the equations.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by MichaelClerx** [comment]
Possibly in test_sbml_ parser you could get the eq from a parsed sbml model without doing the to-myokit-model step? Then just e.g. check its str() representation

5 Jan 2022 16:46:46 David Augustin ***@***.***>:

**Comment by MichaelClerx** [comment]
Thanks @DavAug !

**Pull request merged**